File: client/src/main/java/au/com/codeka/warworlds/game/starfield/scene/TacticalTexture.java
package au.com.codeka.warworlds.game.starfield.scene;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RadialGradient;
import android.graphics.Shader;
import android.opengl.GLES20;
import android.opengl.GLUtils;

import androidx.annotation.Nullable;

import com.google.common.base.Preconditions;

import au.com.codeka.BackgroundRunner;
import au.com.codeka.common.Log;
import au.com.codeka.common.Pair;
import au.com.codeka.common.model.BaseColony;
import au.com.codeka.common.model.BaseFleet;
import au.com.codeka.common.model.BaseStar;
import au.com.codeka.warworlds.model.Empire;
import au.com.codeka.warworlds.model.EmpireManager;
import au.com.codeka.warworlds.model.EmpireShieldManager;
import au.com.codeka.warworlds.model.Sector;
import au.com.codeka.warworlds.model.SectorManager;
import au.com.codeka.warworlds.opengl.Texture;

/**
 * A [Texture] that returns a bitmap to use as a texture for the "tactical" SceneObject that
 * covers a sector.
 */
class TacticalTexture extends Texture {
  private static final Log log = new Log("TacticalTexture");

  // The size of the texture, in pixels, that we'll generate.
  private static final int TEXTURE_SIZE = 128;

  // The radius of the circle, in pixels, that we'll put around each empire.
  private static final int CIRCLE_RADIUS = 30;

  private static final int[] gradientColors = {0, 0};
  private static final float[] gradientStops = {0.1f, 1.0f};

  private static final TacticalTextureLruCache textureCache = new TacticalTextureLruCache(16);

  private final long sectorX;
  private final long sectorY;
  @Nullable
  private Bitmap bitmap;


  private TacticalTexture(long sectorX, long sectorY) {
    this.sectorX = sectorX;
    this.sectorY = sectorY;

    new BackgroundRunner<Bitmap>() {
      @Override
      protected Bitmap doInBackground() {
        return createBitmap(sectorX, sectorY);
      }

      @Override
      protected void onComplete(Bitmap bitmap) {
        TacticalTexture.this.bitmap = bitmap;
      }
    }.execute();
  }

  public void close() {
    if (bitmap != null) {
      bitmap.recycle();
    }
  }

  @Override
  public void bind() {
    if (bitmap != null) {
      setTextureId(createGlTexture());
      bitmap.recycle();
      bitmap = null;
    }
    super.bind();
  }

  private int createGlTexture() {
    Preconditions.checkState(bitmap != null);

    int[] textureHandleBuffer = new int[1];
    GLES20.glGenTextures(1, textureHandleBuffer, 0);
    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textureHandleBuffer[0]);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_LINEAR);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_LINEAR);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_S, GLES20.GL_CLAMP_TO_EDGE);
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_T, GLES20.GL_CLAMP_TO_EDGE);
    GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bitmap, 0);
    return textureHandleBuffer[0];
  }

  public static TacticalTexture create(long sectorX, long sectorY) {
    synchronized(textureCache) {
      Pair<Long, Long> sectorCoord = new Pair<>(sectorX, sectorY);
      TacticalTexture cachedTexture = textureCache.get(sectorCoord);
      if (cachedTexture != null) {
        return cachedTexture;
      }
      TacticalTexture texture = new TacticalTexture(sectorX, sectorY);
      textureCache.put(sectorCoord, texture);
      return texture;
    }
  }

  private static Bitmap createBitmap(long sectorX, long sectorY) {
    log.info("Generating tactical bitmap for %d,%d...", sectorX, sectorY);
    long startTime = System.currentTimeMillis();

    Bitmap bmp = Bitmap.createBitmap(TEXTURE_SIZE, TEXTURE_SIZE, Bitmap.Config.ARGB_8888);
    Canvas canvas = new Canvas(bmp);
    Paint paint = new Paint();
    paint.setStyle(Paint.Style.FILL);

    // We have to look at sectors around this one as well, so that edges look right
    for (long offsetY = -1; offsetY <= 1; offsetY ++) {
    for (long offsetX = -1; offsetX <= 1; offsetX ++) {
        drawCircles(sectorX, sectorY, offsetX, offsetY, canvas, paint);
      }
    }

    long endTime = System.currentTimeMillis();
    log.info("Tactical bitmap for %d,%d generated in %dms", sectorX, sectorY, endTime - startTime);
    return bmp;
  }

  private static void drawCircles(
      long sectorX, long sectorY, long offsetX, long offsetY, Canvas canvas, Paint paint) {
    float scaleFactor = TEXTURE_SIZE / 1024f;
    Sector sector = SectorManager.i.getSector(sectorX, sectorY);
    if (sector == null) {
      return;
    }
    for (BaseStar star : sector.getStars()) {
      float x = (star.getOffsetX() + offsetX * 1024.0f) * scaleFactor;
      float y = (star.getOffsetY() + offsetY * 1024.0f) * scaleFactor;
      float radius = CIRCLE_RADIUS;
      if (x < -radius || x > TEXTURE_SIZE + radius || y < -radius || y > TEXTURE_SIZE + radius) {
        // if it's completely off the bitmap, skip drawing it
        continue;
      }
      int color = 0;
      Integer empireId = getEmpireId(star);
      if (empireId == null) {
        empireId = getFleetOnlyEmpire(star);
        if (empireId != null) {
          color = getShieldColour(empireId);
          color = 0x66ffffff & color;
        }
      } else {
        color = getShieldColour(empireId);
      }

      gradientColors[0] = color;
      gradientColors[1] = color & 0x00ffffff;
      RadialGradient gradient = new RadialGradient(
        x, y, CIRCLE_RADIUS, gradientColors, gradientStops, Shader.TileMode.CLAMP);
      paint.setShader(gradient);
      canvas.drawCircle(x, y, radius, paint);
    }
  }

  private static int getShieldColour(Integer empireId) {
    Empire empire = EmpireManager.i.getEmpire(empireId);
    if (empire == null) {
      return 0;
    }
    return EmpireShieldManager.i.getShieldColour(empire);
  }

  private static Integer getEmpireId(BaseStar star) {
    // if it's a wormhole, the empire is the owner of the wormhole
    BaseStar.WormholeExtra wormholeExtra = star.getWormholeExtra();
    if (wormholeExtra != null) {
      return wormholeExtra.getEmpireID();
    }

    // otherwise, pick the first colony we find to represent the star
    // TODO: what if the star is colonized by more than one empire?
    for (BaseColony colony : star.getColonies()) {
      if (colony.getEmpireKey() != null) {
        return Integer.parseInt(colony.getEmpireKey());
      }
    }

    return null;
  }

  /**
   * If we can't find an empire by colony, look for a fleet. If there's one, we'll display a circle
   * that a little dimmed.
   */
  private static Integer getFleetOnlyEmpire(BaseStar star) {
    // TODO: what if there is more than one fleet? Should try to mix the colors?
    for (BaseFleet fleet : star.getFleets()) {
      if (fleet.getEmpireKey() != null) {
        return Integer.parseInt(fleet.getEmpireKey());
      }
    }

    return null;
  }
}

File: client/src/main/kotlin/au/com/codeka/warworlds/client/game/starfield/scene/TacticalTexture.kt
package au.com.codeka.warworlds.client.game.starfield.scene

import android.graphics.*
import android.opengl.GLES20
import android.opengl.GLUtils
import au.com.codeka.warworlds.client.App
import au.com.codeka.warworlds.client.concurrency.Threads
import au.com.codeka.warworlds.client.game.world.StarManager
import au.com.codeka.warworlds.client.opengl.Texture
import au.com.codeka.warworlds.common.Log
import au.com.codeka.warworlds.common.proto.SectorCoord
import au.com.codeka.warworlds.common.proto.Star
import com.google.common.base.Preconditions
import java.util.*

/**
 * A [Texture] that returns a bitmap to use as a texture for the "tactical" SceneObject that
 * covers a sector.
 */
class TacticalTexture private constructor(private val sectorCoord: SectorCoord) : Texture() {
  private var bitmap: Bitmap? = null

  init {
    App.taskRunner.runTask({ bitmap = createBitmap(sectorCoord) }, Threads.BACKGROUND)
  }

  override fun bind() {
    if (bitmap != null) {
      setTextureId(createGlTexture())
      bitmap?.recycle()
      bitmap = null
    }
    super.bind()
  }

  private fun createGlTexture(): Int {
    Preconditions.checkState(bitmap != null)
    val textureHandleBuffer = IntArray(1)
    GLES20.glGenTextures(1, textureHandleBuffer, 0)
    GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textureHandleBuffer[0])
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MIN_FILTER, GLES20.GL_LINEAR)
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_MAG_FILTER, GLES20.GL_LINEAR)
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_S, GLES20.GL_CLAMP_TO_EDGE)
    GLES20.glTexParameteri(GLES20.GL_TEXTURE_2D, GLES20.GL_TEXTURE_WRAP_T, GLES20.GL_CLAMP_TO_EDGE)
    GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bitmap, 0)
    return textureHandleBuffer[0]
  }

  companion object {
    private val log = Log("TacticalTexture")

    // The size of the texture, in pixels, that we'll generate.
    private const val TEXTURE_SIZE = 128

    // The radius of the circle, in pixels, that we'll put around each empire.
    private const val CIRCLE_RADIUS = 30

    private val gradientColors = intArrayOf(0, 0)
    private val gradientStops = floatArrayOf(0.1f, 1.0f)

    private val textureCache = TacticalTextureLruCache(16)

    fun create(sectorCoord: SectorCoord): TacticalTexture {
      synchronized(textureCache) {
        val cachedTexture = textureCache[sectorCoord]
        if (cachedTexture != null) {
          return cachedTexture
        }
        val texture = TacticalTexture(sectorCoord)
        textureCache.put(sectorCoord, texture)
        return texture
      }
    }

    private fun createBitmap(sectorCoord: SectorCoord): Bitmap {
      log.info("Generating bitmap for ${sectorCoord.x},${sectorCoord.y}...")
      val startTime = System.currentTimeMillis()

      val bmp = Bitmap.createBitmap(TEXTURE_SIZE, TEXTURE_SIZE, Bitmap.Config.ARGB_8888)
      val canvas = Canvas(bmp)
      val paint = Paint()
      paint.style = Paint.Style.FILL

      // We have to look at sectors around this one as well, so that edges look right
      for (offsetY in -1..1) {
        for (offsetX in -1..1) {
          drawCircles(sectorCoord, offsetX, offsetY, canvas, paint)
        }
      }

      val endTime = System.currentTimeMillis()
      log.info("Bitmap for ${sectorCoord.x},${sectorCoord.y} generated in ${endTime - startTime}ms")
      return bmp
    }

    private fun drawCircles(
        sectorCoord: SectorCoord, offsetX: Int, offsetY: Int, canvas: Canvas, paint: Paint) {
      val scaleFactor = TEXTURE_SIZE.toFloat() / 1024f
      StarManager.searchSectorStars(
          SectorCoord(x = sectorCoord.x + offsetX, y = sectorCoord.y + offsetY)).use {
        starCursor ->
          for (star in starCursor) {
            val x = (star.offset_x + offsetX * 1024.0f) * scaleFactor
            val y = (star.offset_y + offsetY * 1024.0f) * scaleFactor
            val radius = CIRCLE_RADIUS.toFloat()
            if (x < -radius || x > TEXTURE_SIZE + radius ||
                y < -radius || y > TEXTURE_SIZE + radius) {
              // if it's completely off the bitmap, skip drawing it
              continue
            }
            var color = 0
            var empireId = getEmpireId(star)
            if (empireId == null) {
              empireId = getFleetOnlyEmpire(star)
              if (empireId != null) {
                color = getShieldColour(empireId)
                color = 0x66ffffff and color
              }
            } else {
              color = getShieldColour(empireId)
            }

            gradientColors[0] = color
            gradientColors[1] = color and 0x00ffffff
            val gradient = RadialGradient(
                x, y, CIRCLE_RADIUS.toFloat(), gradientColors, gradientStops, Shader.TileMode.CLAMP)
            paint.shader = gradient
            canvas.drawCircle(x, y, radius, paint)
          }
        }
    }

    // TODO: this logic is basically cut'n'paste from EmpireRendererHandler on the server.
    private fun getShieldColour(empireId: Long?): Int {
      if (empireId == null) {
        return 0
      }
      val rand = Random(empireId xor 7438274364563846L)
      return -0x1000000 or (
          rand.nextInt(100) + 100 shl 16) or (
          rand.nextInt(100) + 100 shl 8) or
          rand.nextInt(100) + 100
    }

    private fun getEmpireId(star: Star): Long? {
      // if it's a wormhole, the empire is the owner of the wormhole
//    BaseStar.WormholeExtra wormholeExtra = star.getWormholeExtra();
//    if (wormholeExtra != null) {
//      return EmpireManager.i.getEmpire(wormholeExtra.getEmpireID());
//    }

      // otherwise, pick the first colony we find to represent the star
      // TODO: what if the star is colonized by more than one empire?
      for (planet in star.planets) {
        val colony = planet.colony
        if (colony?.empire_id != null) {
          return colony.empire_id
        }
      }

      return null
    }

    /**
     * If we can't find an empire by colony, look for a fleet. If there's one, we'll display a circle
     * that a little dimmed.
     */
    private fun getFleetOnlyEmpire(star: Star): Long? {
      // TODO: what if there is more than one fleet? Should try to mix the colors?
      for (fleet in star.fleets) {
        if (fleet.empire_id != null) {
          return fleet.empire_id
        }
      }
      return null
    }
  }
}